presto-product-tests/src/main/java/com/facebook/presto/tests/SqlCancelTests.java

```diff
@@ -0,0 +1,212 @@
+    {
+        private String sql;
+
+        public SqlExecuteCallable(String sql)
+        {
+            this.sql = sql;
+        }
+
+        public Boolean call()
+        {
+            try {
+                query(sql);
+                return false;
+            }
+            catch (QueryExecutionException e) {
+                return true;
+            }
+        }
+    }
+
+    private class SqlCancelCallable
+            implements Callable<Response>
+    {
+        private String sqlStatement;
+        private long waitTime;
+        private long step;
+
+        public SqlCancelCallable(String sqlStatement, long waitTime, long step)
+        {
+            this.sqlStatement = sqlStatement;
+            this.waitTime = waitTime;
+            this.step = step;
+        }
+
+        public Response call()
+        {
+            long time = 0;
+            while (time < waitTime) {
+                String queryState = (String) query(format("SELECT state from system.runtime.queries" +
+                        " WHERE query = '%s' ORDER BY created DESC LIMIT 1", sqlStatement)).row(0).get(0);
+                if (!queryState.equalsIgnoreCase("running")) {
+                    try {
+                        Thread.sleep(step);
+                    }
+                    catch (InterruptedException e) {
+                        e.printStackTrace();
+                        return null;
+                    }
+                    time = time + step;
+                }
+                else {
+                    String queryId = (String) query(format("SELECT query_id from system.runtime.queries" +
+                            " WHERE query = '%s' ORDER BY created DESC LIMIT 1", sqlStatement)).row(0).get(0);
+                    return (new CancelQuery(queryId, serverAddress)).cancel();
+                }
+            }
+            return null;
+        }
+    }
+
+    private void doCancelTest(String sql)
+    {
+        Response response = null;
+        Boolean isQueryCancelled;
+
+        ExecutorService executor = Executors.newFixedThreadPool(2);
+        Future<Boolean> executeQueryThread = executor.submit(new SqlExecuteCallable(sql));
+        Future<Response> cancelQueryThread = executor.submit(new SqlCancelCallable(sql, 10000, 100));
+
+        try {
+            isQueryCancelled = executeQueryThread.get(2, TimeUnit.MINUTES);
+        }
+        catch (InterruptedException | ExecutionException e) {
+            throw new RuntimeException("error in query execution thread", e);
+        }
+        catch (TimeoutException e) {
+            executeQueryThread.cancel(true);
+            throw new RuntimeException("query execution timelimit exceeded", e);
+        }
+
+        String queryState = (String) query(format("SELECT state from system.runtime.queries" +
+                " WHERE query = '%s' ORDER BY created DESC LIMIT 1", sql)).row(0).get(0);
+        if (!isQueryCancelled || !queryState.equalsIgnoreCase("failed")) {
```

Comment:
Exceptions should only be used for error handling.  For things you want to check in a test, you should use assert statements.  In this case standard testNG assertions will work (https://jitpack.io/com/github/cbeust/testng/master-6.11-gc0bb558-111/javadoc/org/testng/Assert.html)